Fixes #17962 - extend core roles

ares opened the pull request:
Most of this PR is moving permissions definition to plugin so it's executed in right order. Then I added `add_all_permissions_to_default_roles` at the end of plugin registration block that extends core Manager and Viewer role. This bumps Foreman dependency to 1.15.

ares commented:
I highly recommend getting this in before next release. The test failures seems to be related to https://github.com/theforeman/foreman-packaging/pull/1573

ares commented:
Should be fixed in core now [test]

ehelms commented:
@ares can you explain a bit more why we have to pile all the permissions into a single file? You mentioned to ensure correct order, but it's not obvious to me why ordering is required to load in permissions.

jlsherrill commented:
Being in a single file by itself would be better than putting it in plugin.rb IMHO.  plugin.rb is getting quite crowded with lots of little things, sticking all the perms in it makes it a bit unwieldy

ares commented:
I'm not sure how to do it right. If I extract it to a file and I use require, I need to hook into plugin context somehow. The plugin is not registered yet at that point. I could use `eval` but... Or I can create a class that receives plugin context and call permissions on it (see second commit), probably most clean. But IMHO all these solutions are wrong.

Honestly I'd recommend trying to avoid diverging from how other plugins do permissions definition. That's the way Katello can avoid breakage by Foreman changes in future. The way Katello defined permissions so far is a good example.

I understand the list is huge. I'd encourage to raising an issue in Foreman or send a patch to core that would address Katello needs, instead of workarounding it. Maybe by introducing PermissionCreator that would only receive data required to define permissions. Core would make sure it would be defined at the right time.

Anyway, let me know if you find this better and want me to squash.

ares commented:
> @ares can you explain a bit more why we have to pile all the permissions into a single file? You mentioned to ensure correct order, but it's not obvious to me why ordering is required to load in permissions.

Katello used to find the registered plugin and run security_block on it. We need to call `add_all_permissions_to_default_roles` within register block. If we simply required permission definitions in register block, it would not find the plugin since it's registration hasn't finished yet.

Note that plugin register also defines some order, e.g. after_intialize block run just after the register block is evaluated, therefore if some after_initialize code relied on permission to exist, it would fail.

ehelms commented:
@ares Could the require be at the bottom of the plugin definition? I feel like being able to organize your code the way ruby is designed to allow you shouldn't be classified as diverging from other plugins :) Code should just work. If it doesn't, then we will just accept this as is and let's file a bug.

ares commented:
@ehelms I don't think that require helps. Let's assume this is how other plugin defines it.

```
Plugin.register do
  define_permission
  extend_roles_with_defined_permissions
end
```

Not perfect since extend_roles_with_defined_permissions needs to be run after define_permission and both needs to be run withing plugin registration block. But that's the way core currently allows to add permissions.

What Katello [did before the patch](https://github.com/Katello/katello/pull/6703/files#diff-cf74e596d7887011413ee9d06b16b753R94) was something like 

```
Plugin.register do
  ...
end

# in after_initialize that gets run after the plugin is registered
require 'define_permissions_in_external_file'
```

and an external file was doing the definition like this
```
Foreman::Plugin.find(:katello).security_block :capsule_content do
 define_permission
end
```

The main difference is that it defined permissions outside of register block. Note that the register method ensures other stuff is run [before/after the block](https://github.com/theforeman/foreman/blob/develop/app/services/foreman/plugin.rb#L67). So far it worked for Katello but it there was another after initialize code in core that would rely on all permissions being defined, it would not find Katello permissions.

Now the change I'm proposing needs to run `add_all_permissions_to_default_roles` in register block. Therefore I moved the permission registration before it. I can't use require there, since the required file relies on the fact that the plugin is already registered but we'd require it from the middle of registration. As I said, it's not ideal to force plugin to define everything in the register block but that's how other plugins do it. So as a compromise I added a new class that contains the permission registration logic so it is in separate file. The diverging I was referring to is how Katello defined permissions before and how it does it in this PR. It's not a big deal but if there's a bug in the PermissionCreator class, it would fall on Katello devs to fix it. Therefore I'd recommend keep it as close as possible to other plugins.

ares commented:
Is there something I should explain better or I could do to get this moving?

ehelms commented on lib/katello/permission_creator.rb:
```diff
@@ -0,0 +1,359 @@
+class PermissionCreator
```
Can we wrap this in `module Katello` for safety?

ares replied:
good idea :+1:

ehelms commented on lib/katello/engine.rb:
```diff
@@ -90,6 +90,7 @@ def find_assets(args = {})
     # hook so that the resumed Dynflow tasks can rely on everything ready.
     initializer 'katello.register_plugin', :before => :finisher_hook do
       require 'katello/plugin'
+      # extend buildin permissions from core with new actions
```
Minor s/buildin/builtin

ehelms commented:
I think the compromise is a good one and appreciated @ares 

Two minor comments

ares commented:
I updated the PR and squashed commits. Hopefully it will remain green :-)

ares commented:
:green_apple: :-)

jlsherrill commented:
Thanks @ares !!!
